File: 백준/Silver/10164. 격자상의 경로/격자상의 경로.py
import sys
import math

class Pos:
    def __init__(self, y, x):
        self.y = y
        self.x = x

    def __str__(self):
        return 'y : ' + str(self.y) + ', x : ' + str(self.x)

def countWays(positions: list[Pos]):
    ways = 1
    for i in range(1, len(positions)):
        yCnt = positions[i].y - positions[i - 1].y
        xCnt = positions[i].x - positions[i - 1].x
        ways *= math.factorial(yCnt + xCnt) // (math.factorial(yCnt) * math.factorial(xCnt))

    return ways

def main():
    ySize, xSize, via = map(int, sys.stdin.readline().split())
    positions = []

    via -= 1

    positions.append(Pos(0, 0))
    if via > 0:
        positions.append(Pos(via // xSize, via % xSize))
    positions.append(Pos(ySize - 1, xSize - 1))

    # for pos in positions:
    #     print(pos)
    print(countWays(positions))

main()

File: 백준/Silver/10164. 격자상의 경로/격자상의 경로.cc
#include <iostream>
#include <iomanip>
#include <vector>

struct t_pos {
    int y, x;
};

double fact(int n) {
    double ret = 1;

    while (n > 1) {
        ret *= n;
        --n;
    }
    return ret;
}

double countWays(std::vector<t_pos> &positions) {
    double ways = 1;

    for (int i = 1; i < positions.size(); i++) {
        const t_pos &prev = positions[i - 1];
        const t_pos &cur = positions[i];
        const int yCnt = cur.y - prev.y;
        const int xCnt = cur.x - prev.x;

        ways *= fact(yCnt + xCnt) / (fact(yCnt) * fact(xCnt));
    }
    return ways;
}

int main(void) {
    std::cin.tie(0)->sync_with_stdio(0);

    int ySize, xSize, via;
    std::vector<t_pos> positions;

    std::cin >> ySize >> xSize >> via;

    positions.push_back({0, 0});
    if (--via > 0) {
        positions.push_back({via / xSize, via % xSize});
    }
    positions.push_back({ySize - 1, xSize - 1});

    std::cout << std::setprecision(0) << std::fixed << countWays(positions) << '\n';
    return 0;
}
